File: ProjectTracker20cs/ProjectTracker.Library/ProjectList.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Data;
using System.Data.SqlClient;
using Csla;
using Csla.Data;

namespace ProjectTracker.Library
{
  [Serializable()]
  public class ProjectList : 
    ReadOnlyListBase<ProjectList, ProjectList.ProjectInfo>
  {
    #region ProjectInfo Class

    [Serializable()]
    public class ProjectInfo :
      ReadOnlyBase<ProjectInfo>
    {
      #region Business Methods

      private Guid _id;
      private string _name;

      public Guid Id
      {
        get { return _id; }
      }

      public string Name
      {
        get { return _name; }
      }

      protected override object GetIdValue()
      {
        return _id;
      }

      public override string ToString()
      {
        return _name;
      }
      
      #endregion

      #region Constructors

      private ProjectInfo()
      { /* require use of factory methods */ }

      internal ProjectInfo(Guid id, string name)
      {
        _id = id;
        _name = name;
      }
      #endregion
    }

    #endregion

    #region Factory Methods

    /// <summary>
    /// Return a list of all projects.
    /// </summary>
    public static ProjectList GetProjectList()
    {
      return DataPortal.Fetch<ProjectList>(new Criteria());
    }

    /// <summary>
    /// Return a list of projects filtered
    /// by project name.
    /// </summary>
    public static ProjectList GetProjectList(string name)
    {
      return DataPortal.Fetch<ProjectList>
        (new FilteredCriteria(name));
    }

    private ProjectList()
    { /* require use of factory methods */ }

    #endregion

    #region Data Access

    [Serializable()]
    private class Criteria
    { /* no criteria - retrieve all projects */ }

    [Serializable()]
    private class FilteredCriteria
    {
      private string _name;
      public string Name
      {
        get { return _name; }
      }

      public FilteredCriteria(string name)
      {
        _name = name;
      }
    }

    private void DataPortal_Fetch(Criteria criteria)
    {
      // fetch with no filter
      Fetch("");
    }

    private void DataPortal_Fetch(FilteredCriteria criteria)
    {
      Fetch(criteria.Name);
    }

    private void Fetch(string nameFilter)
    {
      using (SqlConnection cn = new SqlConnection(Database.PTrackerConnection))
      {
        cn.Open();
        using (SqlCommand cm = cn.CreateCommand())
        {
          cm.CommandType = CommandType.StoredProcedure;
          cm.CommandText = "getProjects";
          using (SafeDataReader dr = new SafeDataReader(cm.ExecuteReader()))
          {
            IsReadOnly = false;
            while (dr.Read())
            {
              ProjectInfo info = new ProjectInfo(
                dr.GetGuid(0),
                dr.GetString(1));
              // apply filter if necessary
              if ((nameFilter.Length == 0) || (info.Name.IndexOf(nameFilter) == 0))
                this.Add(info);
            }
            IsReadOnly = true;
          }
        }
      }
    }

    #endregion
  }
}

File: ProjectTracker20vb/ProjectTracker.Library/ProjectList.vb
Imports System.Data.SqlClient
Imports CSLA.Data

<Serializable()> _
Public Class ProjectList
  Inherits ReadOnlyListBase(Of ProjectList, ProjectInfo)

#Region " ProjectInfo Class "

  <Serializable()> _
  Public Class ProjectInfo
    Inherits ReadOnlyBase(Of ProjectInfo)

    Private mId As Guid
    Private mName As String

    Public Property Id() As Guid
      Get
        Return mId
      End Get
      Friend Set(ByVal value As Guid)
        mId = value
      End Set
    End Property

    Public Property Name() As String
      Get
        Return mName
      End Get
      Friend Set(ByVal value As String)
        mName = value
      End Set
    End Property

    Protected Overrides Function GetIdValue() As Object
      Return mId
    End Function

    Public Overrides Function ToString() As String
      Return mName
    End Function

    Private Sub New()
      ' require use of factory methods
    End Sub

    Friend Sub New(ByVal id As Guid, ByVal name As String)
      mId = id
      mName = name
    End Sub

  End Class

#End Region

#Region " Factory Methods "

  Public Shared Function GetProjectList() As ProjectList

    Return DataPortal.Fetch(Of ProjectList)(New Criteria)

  End Function

  Public Shared Function GetProjectList(ByVal name As String) As ProjectList

    Return DataPortal.Fetch(Of ProjectList)(New FilteredCriteria(name))

  End Function

  Private Sub New()
    ' require use of factory methods
  End Sub

#End Region

#Region " Data Access "

  <Serializable()> _
Private Class Criteria
    ' no criteria - retrieve all projects
  End Class

  <Serializable()> _
  Private Class FilteredCriteria
    Private mName As String
    Public ReadOnly Property Name() As String
      Get
        Return mName
      End Get
    End Property

    Public Sub New(ByVal name As String)
      mName = name
    End Sub
  End Class

  Private Overloads Sub DataPortal_Fetch(ByVal criteria As Criteria)

    ' fetch with no filter
    Fetch("")

  End Sub

  Private Overloads Sub DataPortal_Fetch(ByVal criteria As FilteredCriteria)

    Fetch(criteria.Name)

  End Sub

  Private Sub Fetch(ByVal nameFilter As String)

    Using cn As New SqlConnection(DataBase.DbConn)
      cn.Open()
      Using cm As SqlCommand = cn.CreateCommand
        With cm
          .CommandType = CommandType.StoredProcedure
          .CommandText = "getProjects"
          Using dr As New SafeDataReader(.ExecuteReader)
            IsReadOnly = False
            While dr.Read()
              Dim info As New ProjectInfo(dr.GetGuid(0), dr.GetString(1))
              ' apply filter if necessary
              If Len(nameFilter) = 0 OrElse info.Name.IndexOf(nameFilter) = 0 Then
                Me.Add(info)
              End If
            End While
            IsReadOnly = True
          End Using
        End With
      End Using
    End Using

  End Sub

#End Region

End Class
